File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* root, bool &res) {
        if(!root) return 0;
        
        int left = dfs(root->left, res);
        int right = dfs(root->right, res);
        
        if(abs(left-right)>1)
            res = false;
        
        return max(left, right) + 1;
    }
    bool isBalanced(TreeNode* root) {
        if(!root) return true;
        
        bool res = true;
        dfs(root, res);
        return res;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true
    }
      
    _, res:= dfs(root)   
    return res
}

func dfs(root *TreeNode) (int, bool) {
    if root == nil {
        return 0, true
    }
    
    left, leftBool := dfs(root.Left)
    right, rightBool := dfs(root.Right)
    
    height := max(left, right)
    if abs(left-right)>1 {
        return height, false
    }
    res := leftBool && rightBool 
    
    return height+1, res
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}